src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +60,146 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
+    int r;
+
+    bf_assert(marsh && cgens);
+
+    // Iterate over chains and rules to get the counters
+    bf_list_foreach (cgens, cgen_node) {
+        struct bf_cgen *cgen = bf_list_node_get_data(cgen_node);
+        struct bf_counter *counter;
+
+        /**
+         * Each chain has 2 counters: policy and error
+         * Each rule has 1 counter regardless of rule->counters.
+         * So we'll add a zero counter for the case where
+         * rule->counters == false.
+        */
+        bf_list_foreach (&cgen->chain->rules, rule_node) {
+            struct bf_rule *rule = bf_list_node_get_data(rule_node);
+            int map_idx = 0;
+
+            r = bf_counter_new(&counter, 0, 0);
+            if (r)
+                return bf_err_r(r, "failed to create counter\n");
+
+            if (rule->counters) {
+                r = bf_cgen_get_counter(cgen, map_idx, counter);
+                if (r < 0)
+                    return bf_err_r(r, "failed to get rule counter\n");
+            }
+
+            r = bf_list_add_tail(&counters, counter);
+            if (r < 0)
+                return bf_err_r(r, "failed to add rule counter to list\n");
+            map_idx++;
+        }
+
+        // Chain's policy counter
+        r = bf_counter_new(&counter, 0, 0);
+        if (r)
+            return bf_err_r(r, "failed to create counter\n");
+
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_POLICY, counter);
+        if (r < 0)
+            return bf_err_r(r, "failed to get policy counter\n");
+
+        r = bf_list_add_tail(&counters, counter);
+        if (r < 0)
+            return bf_err_r(r, "failed to add policy counter to list\n");
+
+        // Chain's error counter
+        r = bf_counter_new(&counter, 0, 0);
+        if (r)
+            return bf_err_r(r, "failed to create counter\n");
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_ERRORS, counter);
+        if (r < 0)
+            return bf_err_r(r, "failed to get error counter\n");
+
+        r = bf_list_add_tail(&counters, counter);
+        if (r < 0)
+            return bf_err_r(r, "failed to add error counter to list\n");
+    }
+
+    r = bf_list_marsh(&counters, marsh);
+    if (r < 0)
+        return bf_err_r(r, "failed to make counters list marsh\n");
+
+    return 0;
+}
+
+static int _bf_cli_get_rules(const struct bf_request *request,
+                             struct bf_response **response)
+{
+    _cleanup_bf_marsh_ struct bf_marsh *chains_marsh = NULL,
+                                       *counter_marsh = NULL,
+                                       *result_marsh = NULL;
+    _clean_bf_list_ bf_list chains = bf_list_default(NULL, bf_chain_marsh);
+    _clean_bf_list_ bf_list _chains = bf_list_default(NULL, bf_chain_marsh);
+    _clean_bf_list_ bf_list cgens = bf_list_default(NULL, NULL);
+    int r;
+
+    // Empty context, nothing to return
+    if (bf_ctx_is_empty())
+        return bf_response_new_success(response, NULL, 0);
```

Comment:
No need, as we iterate over lists, an empty context should be handled properly with the existing code, and create the proper structure for the response.